base/range.jl

```diff
@@ -1480,3 +1480,111 @@ julia> mod(3, 0:2)  # mod(3, 3)
 """
 mod(i::Integer, r::OneTo) = mod1(i, last(r))
 mod(i::Integer, r::AbstractUnitRange{<:Integer}) = mod(i-first(r), length(r)) + first(r)
+
+
+"""
+    LogRange(start, stop, length)
+
+Construct a specialized array whose elements are spaced logarithmically
+between the given endpoints. That is, the ratio of successive elements is
+a constant, calculated from the length.
+
+Like [`LinRange`](@ref), floating-point error may cause it to miss intermediate
+rational values, but the first and last elements should be exactly those provided.
+
+This is similar to `geomspace` in Python. Unlike `PowerRange` in Mathematica,
+you specify the number of elements not the ratio.
+Unlike `logspace` in Python and Matlab, the `start` and `stop` arguments are
+always the first and last elements of the result, not powers applied to some base.
+
+!!! compat "Julia 1.9"
+    This function requires at least Julia 1.9.
+
+# Examples
+```jldoctest
+julia> LogRange(10, 4000, 3)
+3-element LogRange{Float64}:
+ 10.0, 200.0, 4000.0
+
+julia> ans[2] ≈ sqrt(10 * 4000)  # middle element is the geometric mean
+true
+
+julia> LogRange(-62_500, -0.01, 5) |> collect  # endpoints are always exact
+5-element Vector{Float64}:
+ -62500.0
+  -1250.0
+    -25.0
+     -0.49999999999999994
+     -0.01
+
+julia> LogRange(√2, 32, 10)
+10-element LogRange{Float64}:
+ 1.41421, 2.0, 2.82843, 4.0, 5.65685, 8.0, 11.3137, 16.0, 22.6274, 32.0
+
+julia> LogRange(1, 1000, 4) ≈ 10 .^ (0:3)  # was logspace(0,3,4) in Julia < 1.0
+true
+
+julia> LogRange(1, -1+0im, 5) ≈ cis.(LinRange(0, pi, 5))  # complex numbers
+true
+```
+"""
+struct LogRange{T} <: AbstractArray{T,1}
+    start::T
+    stop::T
+    len::Int
+    function LogRange(start::T1, stop::T2, len::Integer) where {T1<:Number, T2<:Number}
+        T = float(promote_type(T1, T2))
+        lo = iszero(start) ? T(NaN) : T(start)
+        hi = iszero(stop) ? T(NaN) : T(stop)
+        if len < 0
+            throw(ArgumentError("LogRange can't have negative length"))
+        elseif len == 0
+            return new{T}(lo, hi, len)
+        elseif len == 1 && start != stop
+            throw(ArgumentError("endpoints of LogRange differ, with length = 1"))
+        end
+        if T <: Real && (start<0) ⊻ (stop<0)
+            throw(DomainError((start, stop),
+                "LogRange will only return complex results if called with a complex argument"))
+        end
+        new{T}(lo, hi, len)
+    end
+end
+
+size(r::LogRange) = (r.len,)
+
+first(r::LogRange) = r.start
+last(r::LogRange) = r.stop
+
+function getindex(r::LogRange, i::Integer)
```

Comment:
I believe you only need to define `getindex(r::LogRange, i::Int)` and the generic machinery will handle rejecting `Bool`s and converting to `Int`.

(see https://docs.julialang.org/en/v1/manual/style-guide/#Handle-excess-argument-diversity-in-the-caller for generic motivation and https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array for this specific case)

Reply:
I don't remember any reason to go via `unsafe_getindex` here, except that I was following `LinRange`. This is largely an `AbstractRange` thing:

```julia
julia> Base.unsafe_getindex |> methods
# 4 methods for generic function "unsafe_getindex" from Base:
 [1] unsafe_getindex(r::LinRange, i::Integer)
     @ range.jl:951
 [2] unsafe_getindex(r::StepRangeLen{T, <:Base.TwicePrecision, <:Base.TwicePrecision}, i::Integer) where T
     @ twiceprecision.jl:479
 [3] unsafe_getindex(r::StepRangeLen{T}, i::Integer) where T
     @ range.jl:939
 [4] unsafe_getindex(A::AbstractArray, I...)  # this method just calls @inbounds getindex()
     @ abstractarray.jl:1291
```

This appears to exist (at least partly?) to circumvent `to_indices` and thus *not* convert indices to `Int`. This allows `BigInt` indexing of a `StepRange` too big for any computer, etc.

Happy to remove it and standardise on `Int`.

Reply:
Done.

Reply:
Un-done -- [f7b2cf8](https://github.com/JuliaLang/julia/pull/39071/commits/f7b2cf8768f9b9a51d21fb05e93a35234838131e) restores an `unsafe_getindex` step, because this separates the logic nicely, and dispatches to a different path for Float64.